matrix_functions.py

```diff
@@ -206,13 +257,88 @@ def matrix_eigenvalue_decomposition(
     return L.to(device=current_device), Q.to(device=current_device)
 
 
+def _approximate_eigenvalues_criterion_below_or_equal_tolerance(
+    A: Tensor, Q: Tensor, tolerance: float
+) -> bool:
+    """Evaluates if a criterion using approximate eigenvalues is below or equal to the tolerance.
+
+    Let Q^T A Q =: B be the estimate of the eigenvalues of the matrix A, where Q is the matrix containing the last computed eigenvectors.
+    The approximate eigenvalues criterion is defined as ||B - diag(B)||_F <= tolerance * ||B||_F.  # TODO: Potentially improve the criterion.
+    The tolerance hyperparameter should therefore be in the interval [0.0, 1.0].
+
+    Args:
+        A (Tensor): The symmetric input matrix.
+        Q (Tensor): An estimate of the eigenvectors of A.
+        tolerance (float): The tolerance for the criterion.
+
+    Returns:
+        bool: Whether the criterion is below or equal to the tolerance
+
+    """
+    squared_approximate_eigenvalues = torch.square_(Q.T @ A @ Q)
+    diagonal_summed = squared_approximate_eigenvalues.diag().sum()
+    off_diagonal_summed = squared_approximate_eigenvalues.fill_diagonal_(0.0).sum()
+    norm = torch.sqrt(diagonal_summed + off_diagonal_summed)
+    off_diagonal_norm = torch.sqrt(off_diagonal_summed)
+    return bool(off_diagonal_norm <= tolerance * norm)
```

Comment:
Wondering if it may be faster to just compute the full norm without filling the diagonal and subtracting? 
```suggestion
    approximate_eigenvalues = Q.T @ A @ Q
    total_norm = torch.linalg.norm(approximate_eigenvalues)
    diagonal_norm = torch.linalg.norm(A.diag())
    off_diagonal_norm = torch.sqrt(A_norm ** 2 - diagonal_norm ** 2)
    return bool(off_diagonal_norm <= tolerance * total_norm)
```

Reply:
It seems to be roughly the same, but your suggestion is slightly faster in most settings. I did some additional benchmarking and the speed-ups for the largest matrices seems to be due to the use of `torch.linalg.norm`.

Using `einsum` is a lot slower for smaller matrices and faster for only the largest tested matrix size. Full results:

![Screenshot 2025-03-18 at 13 52 01](https://github.com/user-attachments/assets/de7db9ec-77df-486b-8f89-041922fdb100)